src/faction_camp.cpp

```diff
@@ -1420,20 +1408,136 @@ void basecamp::craft_construction( npc &p, const std::string &cur_id, const std:
     }
 }
 
-bool talk_function::camp_garage_chop_start( npc &p, const std::string &task )
+/*
+ * All three camp_farm_ functions use the same logic, but passing the npc *comp to this function
+ * doesn't work for reasons I don't understand.  So repeat this stuff in two places: once
+ * farm_description and farm_start, and once for farm_return.
+ */
+static size_t camp_farm_act( const tripoint &omt_tgt, std::string &crops, farm_ops op )
 {
-    std::string dir = camp_direction( task );
-    const tripoint omt_pos = p.global_omt_location();
-    tripoint omt_trg;
-    std::vector<std::pair<std::string, tripoint>> om_expansions = om_building_region( p, 1, true );
-    for( const auto &e : om_expansions ) {
-        if( dir == om_simple_dir( omt_pos, e.second ) ) {
-            omt_trg = e.second;
+    size_t harvestable = 0;
+    size_t plots_empty = 0;
+    size_t plots_plow = 0;
+    bool plow = op & farm_ops::plow;
+    bool plant = op & farm_ops::plant;
+    bool harvest = op & farm_ops::harvest;
+
+    const auto is_dirtmound = []( const tripoint & pos, tinymap & bay1, tinymap & bay2 ) {
```

Comment:
Much of this appears to be repeated in the function below. I suggest extracting it to some functions instead of repeating it over and over again.